File: assets/sample.py
from __future__ import annotations

from dataclasses import dataclass
from hashlib import sha256
from typing import Literal, Protocol


UserRole = Literal["admin", "member", "guest"]


@dataclass(frozen=True)
class User:
    id: str
    name: str
    role: UserRole
    active: bool = True


class UserRepository(Protocol):
    def find_by_id(self, user_id: str) -> User | None: ...

    def save(self, user: User) -> None: ...


def hash_user(user: User) -> str:
    """Return a stable hash for cache keys."""
    payload = f"{user.id}:{user.role}:{user.active}"
    return sha256(payload.encode("utf-8")).hexdigest()


def load_user(repo: UserRepository, user_id: str) -> User | None:
    # Skip lookup for empty ids.
    if not user_id.strip():
        return None

    user = repo.find_by_id(user_id)
    if user and user.active:
        print(f"Loaded {user.name} ({user.role})")

    return user

File: assets/sample.rs
use std::fmt;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum UserRole {
    Admin,
    Member,
    Guest,
}

impl fmt::Display for UserRole {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let label = match self {
            Self::Admin => "admin",
            Self::Member => "member",
            Self::Guest => "guest",
        };
        write!(f, "{label}")
    }
}

#[derive(Clone, Debug)]
pub struct User {
    pub id: String,
    pub name: String,
    pub role: UserRole,
    pub active: bool,
}

pub trait UserRepository {
    fn find_by_id(&self, id: &str) -> Option<User>;
    fn save(&mut self, user: User) -> Result<(), String>;
}

/// Returns a stable cache key for the user record.
pub fn cache_key(user: &User) -> String {
    format!("{}:{}:{}", user.id, user.role, user.active)
}

pub fn load_user(repo: &dyn UserRepository, id: &str) -> Option<User> {
    // Skip lookup for empty ids.
    if id.trim().is_empty() {
        return None;
    }

    let user = repo.find_by_id(id)?;
    if user.active {
        println!("Loaded {} ({})", user.name, user.role);
    }

    Some(user)
}
